src/wallet/wallet2.cpp

```diff
@@ -3136,11 +3136,15 @@ bool wallet2::store_keys(const std::string& keys_file_name, const epee::wipeable
   crypto::chacha20(account_data.data(), account_data.size(), key, keys_file_data.iv, &cipher[0]);
   keys_file_data.account_data = cipher;
 
-  unlock_keys_file();
+  std::string tmp_file_name = keys_file_name + ".new";
   std::string buf;
   r = ::serialization::dump_binary(keys_file_data, buf);
-  r = r && epee::file_io_utils::save_string_to_file(keys_file_name, buf); //and never touch wallet_keys_file again, only read
+  r = r && epee::file_io_utils::save_string_to_file(tmp_file_name, buf);
   CHECK_AND_ASSERT_MES(r, false, "failed to generate wallet keys file " << keys_file_name);
+
+  unlock_keys_file();
+  std::error_code e = tools::replace_file(tmp_file_name, keys_file_name);
+  THROW_WALLET_EXCEPTION_IF(e, error::file_save_error, keys_file_name, e);
```

Comment:
Don't throw before you reacquire the lock.

Reply:
And don't throw after you reacquire the lock. Return false instead.

Reply:
Thanks, fixed.

Reply:
D'oh, didn't see the second comment. Ignore for now.

Reply:
How does it look now?

Reply:
I like it